### PR TITLE
RF06: Allow case sensitive quoted aliasing

### LINKED ISSUE #6030: RF06: Unnecessary Quoted Identifier error when column name needs to be case-sensitive
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

When the column name is quoted so that it's case sensitive, the RF06 rule marks the quotes as unnecessary. 

Example Query:
```
SELECT casechange AS "CaseChange"
FROM foo
```

Fix Output:
```
SELECT casechange AS CaseChange
FROM foo
```

### Expected Behaviour

The rule should not fail.

### Observed Behaviour

```
L:   1 | P:  22 | RF06 | Unnecessary quoted identifier "CaseChange".
                       | [references.quoting]
```

### How to reproduce

<details> <summary> test.sql </summary>

```
SELECT casechange AS "CaseChange"
FROM foo
```

</details>

`sqlfluff lint test.sql --dialect ansi`
`sqlfluff fix test_alias.sql --dialect ansi --rules RF06`

### Dialect

ansi

### Version

```
sqlfluff, version 3.1.0
Python 3.11.7
```

### Configuration

NA

### Are you willing to work on and submit a PR to address the issue?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by aayushr7**
### Brief summary of the change made

Allow quoting of columns when the column name needs to be case sensitive.

- fixes #6030 

### Are there any other side effects of this change that we should be aware of?
No

### Pull Request checklist
- [X] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

File: test/fixtures/rules/std_rule_cases/RF06.yml [hunk 1]
```diff
@@ -120,7 +132,7 @@ test_pass_column_reference_prefer_quoted_tsql:
 
 test_fail_column_reference_prefer_quoted_ansi:
   fail_str: |
-    SELECT 123 AS foo;
+    SELECT foo;
```
Comment (WittierDinosaur) [inline test/fixtures/rules/std_rule_cases/RF06.yml:135]: This test doesn't appear to be doing what it should be any more

Comment (WittierDinosaur) [reply test/fixtures/rules/std_rule_cases/RF06.yml:135]: Or rather, cna we keep this as an alias?

Comment (alanmcruickshank) [reply test/fixtures/rules/std_rule_cases/RF06.yml:135]: Having picked through this PR and the others linked to it - I'm not so sure. I think what's happened is that RF06 has got really confusing. I'm going to write up something more comprehensive.

It doesn't help that the docs for this rule are fairly unhelpful.

Comment (aayushr7) [reply test/fixtures/rules/std_rule_cases/RF06.yml:135]: Hi @WittierDinosaur , this test `test_fail_column_reference_prefer_quoted_ansi` looks correct to me. With respect to the configurations in the test: `prefer_quoted_identifiers: true` the query: 
```
SELECT foo;
```
should fail as the correct query under this configuration would be 
```
SELECT "foo";
```

What do you think?

**Comment by alanmcruickshank** [comment]
@aayushr7 - I've put together a second PR which incorporates a lot of the discussion here and from slack. It's only the RF06 portion for now (the AL09 portion is coming), but I'd love your review on it. Your input has definitely shaped it: https://github.com/sqlfluff/sqlfluff/pull/6173

**Comment by alanmcruickshank** [comment]
I'm going to close this PR now that #6173 is approved which supersedes this one.

**Pull request closed without merging**